isis/src/base/apps/bandnorm/bandnorm.cpp

```diff
@@ -0,0 +1,173 @@
+  static vector<double> normalizer;
+  
+  // function prototypes
+  void getStats(Buffer &in);
+  void normalize(Buffer &in, Buffer &out);
+  void Tokenize(const QString &str,
+                vector<QString> & tokens,
+                const QString &delimiters = " ");
+  
+  void bandnorm(UserInterface &ui) {
+    Cube icube(ui.GetFileName("FROM"), "r");
+    bandnorm(&icube, ui);
+  }
+  
+  
+  void bandnorm(Cube *icube, UserInterface &ui) {
+    // We will be processing by line
+    ProcessByLine p;
+  
+    // Now get the statistics for each band or the entire cube
+    QString avg = ui.GetString("AVERAGE");
+    p.SetInputCube(icube);
+    p.StartProcess(getStats);
+    if(avg == "BAND") {
+      int b = 0;
+      Statistics stats;
+      for(int i = 0; i < (int)average.size(); i++) {
+        if(b == band[i]) {
+          stats.AddData(&average[i], (unsigned int)1);
+        }
+        else {
+          normalizer.push_back(stats.Average());
+          b++;
+          stats.Reset();
+        }
+      }
+      normalizer.push_back(stats.Average());
+    }
+    else if(avg == "PENCIL") {
+      TextFile pencil;
+      pencil.Open(ui.GetFileName("SPECTRUM"));
+      std::cout << pencil.LineCount() << " " << icube->bandCount() << std::endl;
+      if(pencil.LineCount() - 1 < icube->bandCount()) {
+        QString msg = "The spectral pencil file [" + ui.GetAsString("SPECTRUM") +
+                      "] does not contain enough data for all bands.";
+        throw IException(IException::User, msg, _FILEINFO_);
+      }
+      QString st;
+      int column = -1;
+      vector<QString> tokens;
+      pencil.GetLine(st);     //Takes care of title line
+      Tokenize(st, tokens, ", \"-+");
+      if(ui.GetAsString("METHOD") == "number") {
+        column = ui.GetInteger("NUMBER");
+      }
+      else {
+        for(unsigned i = 0; i < tokens.size(); i++) {
+          if(tokens[i] == ui.GetString("NAME")) {
+            column = i;
+        break;
+          }
+        }
+      }
+      if(column < 0  || (unsigned)column > tokens.size()) {
+        QString msg = "The column specified in file [" + ui.GetFileName("SPECTRUM")
+                      + "] was not found.";
+        throw IException(IException::User, msg, _FILEINFO_);
+      }
+      // Add the correct column of data to normalizer
+      for(int i = 0; i < icube->bandCount(); i++) {
+        tokens.clear();
+        pencil.GetLine(st);
+        Tokenize(st, tokens, ", \"");
+        std::cout << "col: " << column << std::endl;
+        std::cout << Isis::IString(tokens[column]).ToDouble() << std::endl;
+        normalizer.push_back(Isis::IString(tokens[column]).ToDouble());
+      }
+    }
+    else {  // avg == "CUBE"
+      Statistics stats;
+      for(int i = 0; i < (int)average.size(); i++) {
+        stats.AddData(&average[i], (unsigned int)1);
+      }
+      for(int b = 0; b < icube->bandCount(); b++) {
+        normalizer.push_back(stats.Average());
+      }
+    }
```

Comment:
I believe this is currently over looked in the converted tests

Reply:
The original tests didn't have this coverage either so it wasn't included since I mostly use the original tests as a guide for what to convert over. That being said, adding the coverage seemed simple enough so I included it.